lib/cache/memory-cache-store.js

```diff
@@ -0,0 +1,365 @@
+        }
+
+        for (const key in currentCacheValue.vary) {
+          if (currentCacheValue.vary[key] !== req.headers[key]) {
+            matches = false
+            break
+          }
+        }
+      }
+
+      if (matches) {
+        value = current
+        break
+      }
+    }
+
+    return value
+  }
+}
+
+class MemoryStoreReadableStream extends Readable {
+  /**
+   * @type {MemoryStoreValue}
+   */
+  #value
+  /**
+   * @type {Buffer[]}
+   */
+  #chunksToSend = []
+
+  /**
+   * @param {MemoryStoreValue} value
+   */
+  constructor (value) {
+    super()
+
+    this.#value = value
+
+    if (value.complete) {
+      this.#chunksToSend = [...this.#value.body, null]
+    } else {
+      value.emitter.on('body', () => {
+        this.#chunksToSend = [...this.#value.body, null]
+      })
+
+      value.emitter.on('error', () => {
+        this.#chunksToSend.push(null)
+      })
+    }
+  }
+
+  get value () {
+    return this.#value.value
+  }
+
+  /**
+   * @param {number} size
+   */
+  _read (size) {
+    if (this.#chunksToSend.length === 0) {
+      return
+    }
+
+    if (size > this.#chunksToSend.length) {
+      size = this.#chunksToSend.length
+    }
+
+    for (let i = 0; i < size; i++) {
+      this.push(this.#chunksToSend.shift())
+    }
+  }
+}
+
+// TODO delete the value if it exceeds maxEntrySize or there's an error
+class MemoryStoreWritableStream extends Writable {
+  /**
+   * @type {MemoryStoreValue}
+   */
+  #value
+  #currentSize = 0
+  #maxEntrySize = 0
+  /**
+   * @type {Buffer[]}
+   */
+  #body = []
+  #deleteValueCallback
+
+  /**
+   * @param {MemoryCacheStore} value
+   * @param {number} maxEntrySize
+   */
+  constructor (value, maxEntrySize, deleteValueCallback) {
+    super()
+    this.#value = value
+    this.#maxEntrySize = maxEntrySize
+    this.#deleteValueCallback = deleteValueCallback
+  }
+
+  get rawTrailers () {
+    return this.#value.value.rawTrailers
+  }
+
+  /**
+   * @param {Buffer[] | undefined} trailers
+   */
+  set rawTrailers (trailers) {
+    this.#value.value.rawTrailers = trailers
+  }
+
+  /**
+   * @param {Buffer} chunk
+   * @param {*} _
+   * @param {() => void} callback
+   */
+  _write (chunk, encoding, callback) {
+    if (typeof chunk === 'string') {
+      chunk = Buffer.from(chunk, encoding)
+    }
+
+    this.#currentSize += chunk.byteLength
+    if (this.#currentSize < this.#maxEntrySize) {
+      this.#body.push(chunk)
+    }
+
+    callback()
+  }
+
+  /**
+   * @param {() => void} callback
+   */
+  _final (callback) {
+    if (this.#currentSize < this.#maxEntrySize) {
+      this.#value.complete = true
+      this.#value.body = this.#body
+
+      this.#value.emitter.emit('body')
+      this.#value.emitter = undefined
+    }
+
+    callback()
+  }
+
+  /**
+   * @param {Error} err
+   * @param {() => void} callback
+   */
+  _destroy (err, callback) {
+    if (err) {
+      this.#value.emitter.emit('error', err)
+      this.#deleteValueCallback()
```

Comment:
This is not needed here. Whoever needs to know about the error should just listen to the error event.